### PR TITLE
[stableswap]: Add checks and tests for scaling factors and extend for multi-asset pools

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #2858 and #2861

## What is the purpose of the change

This PR adds checks and tests for core scaling factor logic. It also extends core pool logic to support scaling factors for multi-asset pools.

Note: 95% of the commit history is past PRs that are already merged (was working off of them in parallel)

## Brief Changelog

- Ensure that it is not possible to reach negative scaling factors by setting scaling factors large enough to flip the sign bit during `uint64` -> `int64` conversions
- Add thorough tests for scaling and descaling logic
- Expand core pool logic to allow for multi-asset scaling factors (instead of just taking the first two in the array)
- Move testing logic from commented out scaling factor tests in `pool_service_test.go` at the keeper layer
- Minor refactor for `solveCfmm` (unrelated but simple enough to tie in with other checks)

## Testing and Verifying

- All tests in this PR are in `pool_test.go`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

File: x/gamm/pool-models/stableswap/amm.go [hunk 1]
```diff
@@ -58,15 +58,17 @@ func cfmmConstantMulti(xReserve, yReserve, uReserve, vSumSquares osmomath.BigDec
...
+	} else if len(remReserves) > 0 {
+		uReserve := osmomath.OneDec()
+		wSumSquares := osmomath.ZeroDec()
+		for _, assetReserve := range remReserves {
+			uReserve = uReserve.Mul(assetReserve)
+			wSumSquares = wSumSquares.Add(assetReserve.Mul(assetReserve))
+		}
+		return solveCFMMBinarySearchMulti(cfmmConstantMulti)(xReserve, yReserve, uReserve, wSumSquares, yIn)
+	} else {
+		panic("invalid input reserves for CFMM solver")
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/amm.go:70]: This case should not be impossible since the previous two check cover all possible range of values of `len(remReserves)`

File: x/gamm/pool-models/stableswap/msgs.go [hunk 2]
```diff
@@ -58,6 +58,14 @@ func (msg MsgCreateStableswapPool) ValidateBasic() error {
 		return types.ErrInvalidScalingFactors
 	}
 
+	// iterates len(msg.InitialPoolLiquidity) < 10 times
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/msgs.go:61]: Can we cover this with test please?

File: x/gamm/pool-models/stableswap/pool.go [hunk 3]
```diff
@@ -123,8 +127,14 @@ func (p Pool) getScaledPoolAmts(denoms ...string) ([]sdk.Dec, error) {
 			return []sdk.Dec{}, fmt.Errorf("denom %s does not exist in pool", denom)
 		}
 		scalingFactor := p.GetScalingFactorByLiquidityIndex(liquidityIndex)
+
+		// should not be possible after pool creation checks but we keep this here as a guardrail
+		if int64(scalingFactor) <= 0 {
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/pool.go:132]: nit / non-blocking: at first sight, feels that this might be an overkill / additional complexity given that we already test for this in the constructor, `ValidateBasic`  and `SetStableSwapScalingFactors`

File: x/gamm/pool-models/stableswap/amm.go [hunk 4]
```diff
@@ -58,15 +58,17 @@ func cfmmConstantMulti(xReserve, yReserve, uReserve, vSumSquares osmomath.BigDec
...
 	}
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/amm.go:71]: We should undo this change back to the old code?

This is an unnecessary else if

File: x/gamm/pool-models/stableswap/msgs.go (hunk 2 above)
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/msgs.go:61]:
```suggestion
```

File: x/gamm/pool-models/stableswap/pool.go [hunk 5]
```diff
@@ -24,8 +24,12 @@ var _ types.PoolI = &Pool{}
...
 		scalingFactors = []uint64{1, 1}
-	} else if scalingFactors[0] == 0 || scalingFactors[1] == 0 {
-		return Pool{}, types.ErrInvalidStableswapScalingFactors
+	}
+
+	for _, scalingFactor := range scalingFactors {
+		if scalingFactor == 0 || int64(scalingFactor) <= 0 {
+			return Pool{}, types.ErrInvalidStableswapScalingFactors
+		}
 	}
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/pool.go:33]: This is reused, lets extract this to a function in stableswap for validate scaling factors

File: x/gamm/pool-models/stableswap/pool.go [hunk 6]
```diff
@@ -299,15 +315,16 @@ func (p *Pool) PokePool(blockTime time.Time) {}
 // It should only be able to be successfully called by the pool's ScalingFactorGovernor
 // TODO: move commented test for this function from x/gamm/keeper/pool_service_test.go once a pool_test.go file has been created for stableswap
 func (p *Pool) SetStableSwapScalingFactors(ctx sdk.Context, scalingFactors []uint64, scalingFactorGovernor string) error {
-	if scalingFactorGovernor != p.ScalingFactorGovernor {
+	if scalingFactorGovernor != p.ScalingFactorGovernor || len(scalingFactors) != p.PoolLiquidity.Len() {
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/pool.go:318]: Non-blocking for this PR:

I think checking for governor here is not compatible with the abstraction.

I would like to suggest moving this governor check up to the caller. Then, we can also reuse `SetStableSwapScalingFactors` method in the constructor to reduce code duplication for checking scaling factor overflow

File: x/gamm/pool-models/stableswap/pool.go [hunk 7]
```diff
@@ -123,8 +127,14 @@ func (p Pool) getScaledPoolAmts(denoms ...string) ([]sdk.Dec, error) {
 			return []sdk.Dec{}, fmt.Errorf("denom %s does not exist in pool", denom)
 		}
 		scalingFactor := p.GetScalingFactorByLiquidityIndex(liquidityIndex)
+
+		// should not be possible after pool creation checks but we keep this here as a guardrail
+		if int64(scalingFactor) <= 0 {
+			panic("invalid scaling factor encountered when scaling pool assets")
+		}
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/pool.go:134]: Not sure how I feel about these being everywhere, but if you want this lets add it into the get method, since the property is that this should never get into state.

File: x/gamm/pool-models/stableswap/pool.go [hunk 8]
```diff
@@ -299,15 +315,16 @@ func (p *Pool) PokePool(blockTime time.Time) {}
 // It should only be able to be successfully called by the pool's ScalingFactorGovernor
 // TODO: move commented test for this function from x/gamm/keeper/pool_service_test.go once a pool_test.go file has been created for stableswap
 func (p *Pool) SetStableSwapScalingFactors(ctx sdk.Context, scalingFactors []uint64, scalingFactorGovernor string) error {
-	if scalingFactorGovernor != p.ScalingFactorGovernor {
+	if scalingFactorGovernor != p.ScalingFactorGovernor || len(scalingFactors) != p.PoolLiquidity.Len() {
 		return types.ErrNotScalingFactorGovernor
 	}
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/pool.go:320]: Err should be a different error message and therefore a second check

File: x/gamm/pool-models/stableswap/pool.go [hunk 9]
```diff
@@ -299,15 +315,16 @@ func (p *Pool) PokePool(blockTime time.Time) {}
...
 
-	if len(scalingFactors) != p.PoolLiquidity.Len() {
-		return types.ErrInvalidStableswapScalingFactors
+	for _, scalingFactor := range scalingFactors {
+		if int64(scalingFactor) <= 0 {
+			return types.ErrInvalidStableswapScalingFactors
+		}
 	}
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/pool.go:326]: Another point to reuse the scaling factor validation method

**Review by ValarDragon — APPROVED** [review]
LGTM once my comments are addressed

File: x/gamm/pool-models/stableswap/pool.go (hunk 6 above)
Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/pool.go:318]: The argument should be `sender` here, and this is the entry point of the `SetStableSwapScalingFactors` function right?

File: x/gamm/pool-models/stableswap/msgs.go (hunk 2 above)
Comment (AlpinYukseloglu) [reply x/gamm/pool-models/stableswap/msgs.go:61]: This was a note that I intended to remove, not sure why it's still there (the number of iterations is capped at the max number of pool assets by definition)

File: x/gamm/pool-models/stableswap/amm.go (hunk 4 above)
Comment (AlpinYukseloglu) [reply x/gamm/pool-models/stableswap/amm.go:71]: Reverted

File: x/gamm/pool-models/stableswap/amm.go (hunk 1 above)
Comment (AlpinYukseloglu) [reply x/gamm/pool-models/stableswap/amm.go:70]: Reverted

File: x/gamm/pool-models/stableswap/pool.go (hunk 5 above)
Comment (AlpinYukseloglu) [reply x/gamm/pool-models/stableswap/pool.go]: Great point – extracted into `validateScalingFactors` function

File: x/gamm/pool-models/stableswap/pool.go (hunk 8 above)
Comment (AlpinYukseloglu) [reply x/gamm/pool-models/stableswap/pool.go]: Check moved into `validateScalingFactors` so this should be addressed

File: x/gamm/pool-models/stableswap/pool.go (hunk 3 above)
Comment (AlpinYukseloglu) [reply x/gamm/pool-models/stableswap/pool.go]: I ended up removing this check and the one in `getDescaledPoolAmt` to reduce complexity as all cases should be covered by higher level checks

**Comment by AlpinYukseloglu** [comment]
Thanks for the thorough review! All comments should be addressed

**Review by p0mvn — APPROVED** [review]

**Review by alexanderbez — APPROVED** [review]
